File: Unity/AI_FSM/Assets/Script/State.cs
using System;
using System.Collections.Generic;
using UnityEngine;

public abstract class State : ScriptableObject
{
    [SerializeField] List<Transition> transitions = null;
    [field:SerializeField, HideInInspector] public FSM CurrentFSM { get; private set; }

    protected List<Transition> runningTransitions = new();

    public virtual void Enter(FSM _owner)
    {
        CurrentFSM = _owner;
        InitTransitions();
        Debug.Log("Enter in " + GetType().FullName);
    }

    public virtual void Update()
    {
        CheckForValidTransition();
        Debug.Log("Update in " + GetType().FullName);
    }

    public virtual void Exit()
    {
        Debug.Log("Exit in " + GetType().FullName);
    }

    protected virtual void InitTransitions()
    {
        for (int i = 0; i < transitions.Count; i++)
        {
            if (!transitions[i])
                continue;
            Transition _newTransition = Instantiate(transitions[i]);
            _newTransition.InitTransition(CurrentFSM);
            runningTransitions.Add(_newTransition);
        }
    }

    protected void CheckForValidTransition()
    {
        for (int i = 0; i < runningTransitions.Count; i++)
        {
            if (runningTransitions[i] && runningTransitions[i].IsValidTransition)
            {
                Exit();
                CurrentFSM.SetNextState(runningTransitions[i].NextState);
                return;
            }
        }
    }
}

File: Unreal/IA_FSM/Source/IA_FSM/Private/FSM/State.cpp
// Fill out your copyright notice in the Description page of Project Settings.


#include "FSM/State.h"
#include "FSM/FSMObject.h"
#include "FSM/Transition/WaitProjectilDestroyTransition.h"
#include "FSM/BasicState/ShootState.h"
#include "Kismet/KismetSystemLibrary.h"

void UState::Enter(UFSMObject* _owner)
{
	currentFSM = _owner;
	InitTransitions();
	UKismetSystemLibrary::PrintString(this, "Enter in state " + GetName(), true, true, FColor::Green);
}

void UState::Update()
{
	CheckForValidTransition();
	UKismetSystemLibrary::PrintString(this, "Update in state " + GetName(), true, true, FColor::Yellow, GetWorld()->DeltaTimeSeconds);
}

void UState::Exit()
{
	UKismetSystemLibrary::PrintString(this, "Exit in state " + GetName(), true, true, FColor::Red);
	//Destroy
}

void UState::InitTransitions()
{
	for (TSubclassOf<UTransition> _transition : transitions)
	{
		if (!_transition)
			continue;
		UTransition* _newTransition = NewObject<UTransition>(this, _transition);
		_newTransition->InitTransition();
		runningTransitions.Add(_newTransition);
	}
}

void UState::CheckForValidTransition()
{
	for (TObjectPtr<UTransition> _transition : runningTransitions)
	{
		if (_transition->IsValidTransition())
		{
			Exit();
			currentFSM->SetNextState(_transition->GetNextState());
			return;
		}
	}
}
